packages/dom/src/document.rs

```diff
@@ -522,6 +538,23 @@ impl Document {
     pub fn set_document(&mut self, _content: String) {}
 
     pub fn add_element(&mut self) {}
+
+    pub fn visit<F>(&self, mut visit: F)
+    where
+        F: FnMut(usize, &Node),
+    {
+        let mut stack = VecDeque::new();
+        stack.push_front(0);
+
+        while let Some(node_key) = stack.pop_back() {
+            let node = &self.nodes[node_key];
+            visit(node_key, node);
+
+            for &child_key in &node.children {
+                stack.push_front(child_key);
+            }
+        }
+    }
 }
```

Comment:
Not sure about this (but not totally against it either). You could potentially use the pattern where you `mem::take` the children array before recursing into the children, and then put it back afterwards. That's not exactly ideal either though. Perhaps this pattern is better?

Reply:
Hmm I'm not sure I understand, so we'd be clearing the `children` field while visiting a node's children? I can see that preventing cycles if that's what you're thinking